extra_data/components.py

```diff
@@ -90,6 +89,32 @@ def _check_pulse_selection(pulses):
     return type(pulses)(val)
 
 
+def _select_pulse_ids(pulses, data_pulse_ids):
+    """Select pulses by ID across a chunk of trains
+
+    Returns a boolean array of which entries in data_pulse_ids match.
+    """
+    if isinstance(pulses.value, slice):
+        s = pulses.value
+        desired = np.arange(s.start, s.stop, step=s.step, dtype=np.uint64)
+    else:
+        desired = pulses.value
```

Comment:
I'm 99% sure the `else` branch can only be a `by_index`/`by_id` object, but I wonder if an additional check might prevent cryptic exceptions.

Reply:
Yes, that makes sense.

Reply:
Hmmm, in fact the else branch is something like a list or an array. Presumably, anything which numpy can convert into an array. So it gets a bit more complicated, e.g. do we allow something that converts to an array of floats? :thinking:

Reply:
How so? I tried `isinstance(by_index, slice)` and it's not `True`, and if I traced the code paths correctly whatever is passed *should* be converted to `by_index`/`by_id` if it's not a `slice`.

Reply:
The isinstance check here is for `pulses.value`, i.e. the thing inside a `by_id` object. We could indeed do an extra check where this is called, though.